### ISSUE TITLE
How should SIGPIPE signals be handled? Should it just be ignored on a webserver?

### ISSUE THREAD

**Issue by SimonSimCity** [nodejs/help#2126]
Following the table at https://en.wikipedia.org/wiki/Signal_(IPC)#POSIX_signals I created a bunch of listeners on my application which (among others) would terminate my application when receiving the signal `SIGPIPE`.

After running this application (a webserver) in production, I noticed that the system is going down once in a while. After debugging and the use of `strace`, I was able to confirm that my nodejs process seems to send this signal to itself:

```
16:13:40.525616 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1, si_uid=0} ---
19:09:05.458178 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1, si_uid=0} ---
21:20:05.309759 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1, si_uid=0} ---
21:20:05.310039 --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=1, si_uid=0} ---
```

-> Running node in a docker container, the nodejs process is on pid 1

How can I tell why I receive these signals, and what should I do about them? It doesn't feel good to avoid them if I do not know the exact reason where they're raised and I can rest ensured that I don't need to run an action.

In https://github.com/nodejs/help/issues/1110 @bnoordhuis was talking about a trade-off - would be nice if you can elaborate on this. In the issue I raised earlier in the node-repository (https://github.com/nodejs/node/issues/29165) you mentioned `a TCP socket that was hung up by the remote end` - are there other scenarios?

On https://github.com/tapjs/signal-exit/issues/19#issuecomment-173417833 a user was also expecting a bug in node since it only occurs _sometimes_, and to me it also looks very random on which time/situation I receive one of these signals. Sometimes often on a higher machine load, sometimes it feels like something else is the reason.

**Comment by bnoordhuis** [comment]
The trade-off is that SIGPIPE by default terminates the receiving process. Since reads and writes on non-file objects generate them, Node.js has to either ignore or catch them.

It ignores them because there is no point in catching them. What would it do with the info? Broken connections are already dealt with in libuv.

One thing Node.js could perhaps improve is to restore its SIG_IGN behavior when the last `process.on('SIGPIPE')` listener is removed.

One snag is that right now that can't be done entirely race-free. Libuv restores the signal disposition to SIG_DFL, node.js then has to switch it to SIG_IGN, but that means there is a time window where a SIGPIPE will terminate the process.

I've opened https://github.com/libuv/libuv/issues/2435 to discuss whether that can be addressed in libuv.

**Comment by SimonSimCity** [comment]
Saying that this _can't be done entirely race-free_ is answer enough for me and gives me a good stomach feeling when ignoring these events.

The crashing is only an issue if I've already added a listener for `SIGPIPE` and want to remove it, right? I've registered a custom handler which just logs it out to the console - no action. Means, when removing this listener of mine, it should work nicely and ignore it again.
